### PR TITLE
[SPARK-20341][SQL] Support BigInt's value that does not fit in long value range

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

This PR avoids an exception in the case where `scala.math.BigInt` has a value that does not fit into long value range (e.g. `Long.MAX_VALUE+1`). When we run the following code by using the current Spark, the following exception is thrown.

This PR keeps the value using `BigDecimal` if we detect such an overflow case by catching `ArithmeticException`.

Sample program:
```
case class BigIntWrapper(value:scala.math.BigInt)```
spark.createDataset(BigIntWrapper(scala.math.BigInt("10000000000000000002"))::Nil).show
```
Exception:
```
Error while encoding: java.lang.ArithmeticException: BigInteger out of long range
staticinvoke(class org.apache.spark.sql.types.Decimal$, DecimalType(38,0), apply, assertnotnull(assertnotnull(input[0, org.apache.spark.sql.BigIntWrapper, true])).value, true) AS value#0
java.lang.RuntimeException: Error while encoding: java.lang.ArithmeticException: BigInteger out of long range
staticinvoke(class org.apache.spark.sql.types.Decimal$, DecimalType(38,0), apply, assertnotnull(assertnotnull(input[0, org.apache.spark.sql.BigIntWrapper, true])).value, true) AS value#0
	at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.toRow(ExpressionEncoder.scala:290)
	at org.apache.spark.sql.SparkSession$$anonfun$2.apply(SparkSession.scala:454)
	at org.apache.spark.sql.SparkSession$$anonfun$2.apply(SparkSession.scala:454)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:285)
	at org.apache.spark.sql.SparkSession.createDataset(SparkSession.scala:454)
	at org.apache.spark.sql.Agg$$anonfun$18.apply$mcV$sp(MySuite.scala:192)
	at org.apache.spark.sql.Agg$$anonfun$18.apply(MySuite.scala:192)
	at org.apache.spark.sql.Agg$$anonfun$18.apply(MySuite.scala:192)
	at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
	at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
	at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:68)
	at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:163)
	at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
	at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
	at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:175)
...
Caused by: java.lang.ArithmeticException: BigInteger out of long range
	at java.math.BigInteger.longValueExact(BigInteger.java:4531)
	at org.apache.spark.sql.types.Decimal.set(Decimal.scala:140)
	at org.apache.spark.sql.types.Decimal$.apply(Decimal.scala:434)
	at org.apache.spark.sql.types.Decimal.apply(Decimal.scala)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.toRow(ExpressionEncoder.scala:287)
	... 59 more
```

## How was this patch tested?

Add new test suite into `DecimalSuite`

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 1]
```diff
@@ -135,11 +135,17 @@ final class Decimal extends Ordered[Decimal] with Serializable {
    * Set this Decimal to the given BigInteger value. Will have precision 38 and scale 0.
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: This line needs an update

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 2]
```diff
@@ -135,11 +135,17 @@ final class Decimal extends Ordered[Decimal] with Serializable {
...
+    try {
+      this.decimalVal = null
+      this.longVal = bigintval.longValueExact()
+      this._precision = DecimalType.MAX_PRECISION
+      this._scale = 0
+      this
+    } catch {
+      case _: ArithmeticException =>
+        set(BigDecimal(bigintval))
+        this
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: This line is not needed.

**Review by gatorsmile — COMMENTED** [review]
What if we simply use the the set function for `BigDecimal`?

```Scala
  def set(bigintval: BigInteger): Decimal = set(BigDecimal(bigintval))
```

**Comment by kiszk** [comment]
To always use `BigDecimal` can make code simple. However, it always create and keep an additional object (i.e. `BigDecimal`) even if value can be fit into long value range.
This is a design choice: runtime efficiency with longer code v.s. simpler code with runtime inefficiency.

**Comment by gatorsmile** [comment]
Ok, could you write a comment in the code to explain it?

**Comment by kiszk** [comment]
Thanks, done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 3]
```diff
@@ -132,14 +132,22 @@ final class Decimal extends Ordered[Decimal] with Serializable {
   }
 
   /**
-   * Set this Decimal to the given BigInteger value. Will have precision 38 and scale 0.
+   * Set this Decimal to the given BigInteger value. Will have precision 38 and scale 0
+   * if value is fit into long value range. Otherwise, use BigDecimal
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: How about
> If the value is not in the range of long, convert it to BigDecimal and the precision and scale is based on the converted value.

**Comment by gatorsmile** [comment]
LGTM except one comment. 

cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala (hunk 3 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: Thanks, updated.

**Comment by cloud-fan** [comment]
LGTM, merging to master/2.2!

**Comment by cloud-fan** [comment]
it conflicts with branch 2.1, can you send a new PR? thanks!

**Pull request closed without merging**